src/threading/TaskScheduler.cpp

```diff
@@ -18,62 +19,231 @@
 
 #include "TaskScheduler.hpp"
 
+#include <algorithm>
+#include <atomic>
+#include <iomanip>
+#include <iostream>
+#include <map>
+#include <mutex>
+#include <system_error>
+
+#include "../dsl/word/MainThread.hpp"
+#include "../util/update_current_thread_priority.hpp"
+
 namespace NUClear {
 namespace threading {
 
-    void TaskScheduler::shutdown() {
-        {
-            const std::lock_guard<std::mutex> lock(mutex);
-            running = false;
+    bool TaskScheduler::is_runnable(const std::unique_ptr<ReactionTask>& task, const uint64_t& pool_id) {
+
+        // Task can run if it is meant to run on the current thread pool
+        const bool correct_pool = pool_id == task->thread_pool_descriptor.pool_id;
+
+        // Task can run if the group it belongs to has spare threads
+        if (groups.at(task->group_descriptor.group_id) < task->group_descriptor.thread_count && correct_pool) {
+            // This task is about to run in this group, increase the number of active tasks in the group
+            groups.at(task->group_descriptor.group_id)++;
+            return true;
+        }
+
+        return false;
+    }
+
+    void TaskScheduler::run_task(std::unique_ptr<ReactionTask>&& task) {
+        if (task) {
+            task->run();
+
+            // This task is no longer running, decrease the number of active tasks in the group
+            /* mutex scope */ {
+                const std::lock_guard<std::mutex> group_lock(queue_mutex);
+                groups.at(task->group_descriptor.group_id)--;
+            }
+        }
+    }
+
+    void TaskScheduler::pool_func(const util::ThreadPoolDescriptor& pool) {
+        while (running.load() || !queue.at(pool.pool_id).empty()) {
+            // Wait at a high (but not realtime) priority to reduce latency
+            // for picking up a new task
+            update_current_thread_priority(1000);
+
+            run_task(get_task(pool.pool_id));
+        }
+    }
+
+    TaskScheduler::TaskScheduler() {
+        // Setup everything (thread pool, task queue, mutex, condition variable) for the main thread here
+        pools[util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID] =
+            util::ThreadPoolDescriptor{util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID, 1};
+        pool_map[std::this_thread::get_id()] = util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID;
+        queue.emplace(util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID, std::vector<std::unique_ptr<ReactionTask>>{});
```

Comment:
Did you still want to go back to a single queue or split queue work fine

Reply:
I think split queue is working fine? Unless you think this may have some impact on threads not joining properly?